pod/src/list.rs

```diff
@@ -0,0 +1,334 @@
+/// 1.  **Length**: A `u32` value (`PodU32`) at the beginning of the buffer,
+///     indicating the number of currently active elements in the collection.
+/// 2.  **Data**: The remaining part of the buffer, which is treated as a slice
+///     of `T` elements. The capacity of the collection is the number of `T`
+///     elements that can fit into this data portion.
+pub struct PodList<'data, T: Pod> {
+    length: &'data mut PodU32,
+    data: &'data mut [T],
+    max_length: usize,
+}
+
+impl<'data, T: Pod> PodList<'data, T> {
+    /// Unpack the mutable buffer into a mutable slice, with the option to
+    /// initialize the data
+    fn unpack_internal<'a>(data: &'a mut [u8], init: bool) -> Result<Self, ProgramError>
+    where
+        'a: 'data,
+    {
+        if data.len() < LENGTH_SIZE {
+            return Err(PodSliceError::BufferTooSmall.into());
+        }
+        let (length, data) = data.split_at_mut(LENGTH_SIZE);
+        let length = pod_from_bytes_mut::<PodU32>(length)?;
+        if init {
+            *length = 0.into();
+        }
+        let max_length = max_len_for_type::<T>(data.len(), u32::from(*length) as usize)?;
+        let data = pod_slice_from_bytes_mut(data)?;
+        Ok(Self {
+            length,
+            data,
+            max_length,
+        })
+    }
+
+    /// Unpack the mutable buffer into a mutable slice
+    pub fn unpack<'a>(data: &'a mut [u8]) -> Result<Self, ProgramError>
+    where
+        'a: 'data,
+    {
+        Self::unpack_internal(data, /* init */ false)
+    }
+
+    /// Unpack the mutable buffer into a mutable slice, and initialize the
+    /// slice to 0-length
+    pub fn init<'a>(data: &'a mut [u8]) -> Result<Self, ProgramError>
+    where
+        'a: 'data,
+    {
+        Self::unpack_internal(data, /* init */ true)
+    }
+
+    /// Add another item to the slice
+    pub fn push(&mut self, t: T) -> Result<(), ProgramError> {
+        let length = u32::from(*self.length);
+        if length as usize == self.max_length {
+            Err(PodSliceError::BufferTooSmall.into())
+        } else {
+            self.data[length as usize] = t;
+            *self.length = length.saturating_add(1).into();
+            Ok(())
+        }
+    }
+
+    /// Remove and return the element at `index`, shifting all later
+    /// elements one position to the left.
+    pub fn remove_at(&mut self, index: usize) -> Result<T, ProgramError> {
+        let len = u32::from(*self.length) as usize;
+        if index >= len {
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        let removed_item = self.data[index];
+
+        // Move the tail left by one
+        let tail_start = index
+            .checked_add(1)
+            .ok_or(ProgramError::ArithmeticOverflow)?;
+        self.data.copy_within(tail_start..len, index);
+
+        // Zero-fill the now-unused slot at the end
+        let last = len.checked_sub(1).ok_or(ProgramError::ArithmeticOverflow)?;
```

Comment:
Since the  `len` is already validated, you might not need to do a `checked_sub`.

Reply:
Complaints from the clippy gods on arithmetic side-effects. Either can clippy ignore or do a saturating_sub to appease it.